// frontend/app/integrations/page.tsx
"use client";

import { Card, CardContent, CardDescription, CardHeader, CardTitle } from "@/components/ui/card";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import { 
  MessageSquare, 
  GitGraph, 
  ExternalLink,
  RefreshCw,
  Bell
} from "lucide-react";
import { useEffect, useState } from "react";
import { api } from "@/lib/api";

const FUTURE_INTEGRATIONS = [
  { id: "datadog", name: "Datadog", desc: "Export metrics to Datadog dashboards." },
  { id: "pagerduty", name: "PagerDuty", desc: "Automated incident routing." },
  { id: "discord", name: "Discord", desc: "Rollout alerts to Discord channels." },
  { id: "kubernetes", name: "Kubernetes", desc: "Sync rollout state with K8s." },
  { id: "grafana", name: "Grafana", desc: "Native Grafana dashboards." },
  { id: "launchdarkly", name: "LaunchDarkly", desc: "Migration and sync tool." },
  { id: "sentry", name: "Sentry", desc: "Error tracking correlation." }
];

const IntegrationsPage = () => {
  const [loading, setLoading] = useState(false);
  const [lastSync, setLastSync] = useState("Never");
  const [isHealthy, setIsHealthy] = useState(false);

  const syncIntegrations = async () => {
    setLoading(true);
    try {
      const [health] = await Promise.all([
        api.getHealth(),
        api.getDashboard(),
      ]);
      setIsHealthy(Boolean(health?.environment_ready));
      setLastSync(new Date().toLocaleTimeString());
    } catch (error) {
      console.error("Failed to sync integration status:", error);
      setIsHealthy(false);
      setLastSync("Sync failed");
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    const initialSync = setTimeout(() => {
      void syncIntegrations();
    }, 0);
    return () => clearTimeout(initialSync);
  }, []);

  const integrations = [
    {
      id: "slack",
      name: "Slack",
      description: "Automated rollout notifications and approval workflows.",
      icon: MessageSquare,
      status: isHealthy ? "connected" : "disconnected",
      details: isHealthy ? "Alerts can be routed through backend automation" : "Backend unavailable",
      lastSync
    },
    {
      id: "github",
      name: "GitHub",
      description: "Trigger rollouts from PRs and sync deployment status.",
      icon: GitGraph,
      status: "disconnected",
      details: "Requires OAuth",
      lastSync
    }
  ];

  return (
    <div className="flex-1 space-y-8 p-8 max-w-7xl">
      <div className="flex items-center justify-between">
        <div>
          <h1 className="text-3xl font-bold tracking-tight text-glow">Integrations</h1>
          <p className="text-muted-foreground mt-1">Connect observability, communication, and infrastructure tools.</p>
        </div>
        <Button variant="outline" size="sm" className="rounded-full bg-background/50 backdrop-blur-sm border-border/50" onClick={syncIntegrations} disabled={loading}>
          <RefreshCw className={`mr-2 h-4 w-4 ${loading ? "animate-spin" : ""}`} />
          {loading ? "Syncing..." : "Sync All"}
        </Button>
      </div>

      <div className="grid gap-6 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4">
        {integrations.map((item) => (
          <Card key={item.id} className="glassy-card hover:-translate-y-1 transition-transform duration-300">
            <CardHeader className="pb-3">
              <div className="flex items-center justify-between">
                <div className="flex h-10 w-10 items-center justify-center rounded-lg bg-muted/50">
                  <item.icon className="h-6 w-6 text-primary" />
                </div>
                <Badge 
                  variant={item.status === "connected" ? "default" : "secondary"}
                  className={item.status === "connected" ? "bg-green-500/10 text-green-600 hover:bg-green-500/10" : ""}
                >
                  {item.status === "connected" ? "Connected" : "Disconnected"}
                </Badge>
              </div>
              <CardTitle className="mt-4">{item.name}</CardTitle>
              <CardDescription>{item.description}</CardDescription>
            </CardHeader>
            <CardContent>
              <div className="space-y-4">
                <div className="flex items-center justify-between text-sm">
                  <span className="text-muted-foreground">Status</span>
                  <span className="font-medium">{item.details}</span>
                </div>
                <div className="flex items-center justify-between text-sm">
                  <span className="text-muted-foreground">Last Sync</span>
                  <span className="text-xs">{item.lastSync}</span>
                </div>
                <div className="pt-4 flex gap-2">
                  <Button variant="outline" className="w-full text-xs h-8 rounded-full">
                    Configure
                  </Button>
                  <Button variant="ghost" size="icon" className="h-8 w-8 rounded-full">
                    <ExternalLink className="h-4 w-4" />
                  </Button>
                </div>
              </div>
            </CardContent>
          </Card>
        ))}

        {/* Future Integrations */}
        {FUTURE_INTEGRATIONS.map((item) => (
          <Card key={item.id} className="border border-border/30 bg-card/20 opacity-60 grayscale-[50%] transition-all hover:grayscale-0 hover:opacity-100 cursor-not-allowed">
            <CardHeader className="pb-3">
              <div className="flex items-center justify-between">
                <div className="flex h-10 w-10 items-center justify-center rounded-lg bg-muted/50 border border-border/50">
                  <ExternalLink className="h-5 w-5 text-muted-foreground" />
                </div>
                <Badge variant="outline" className="text-[10px] uppercase tracking-wider bg-background/50 text-muted-foreground">
                  Coming Soon
                </Badge>
              </div>
              <CardTitle className="mt-4 text-muted-foreground">{item.name}</CardTitle>
              <CardDescription className="text-xs">{item.desc}</CardDescription>
            </CardHeader>
          </Card>
        ))}
      </div>

      <Card className="glassy-card border-primary/20 bg-primary/5">
        <CardHeader>
          <CardTitle className="flex items-center gap-2 text-primary">
            <Bell className="h-5 w-5" />
            Webhook Endpoints
          </CardTitle>
          <CardDescription>Receive automated alerts from your own infrastructure.</CardDescription>
        </CardHeader>
        <CardContent className="space-y-4">
          <div className="rounded-lg border border-primary/20 bg-background/50 p-4 font-mono text-xs flex justify-between items-center group">
             <span className="text-primary truncate">GET {(api.getApiBaseUrl() || "http://localhost:8000")}/monitoring/alerts</span>
             <Badge variant="outline" className="opacity-0 group-hover:opacity-100 transition-opacity bg-primary/10 text-primary border-primary/30">Copy</Badge>
          </div>
          
          <div className="flex items-center justify-between border-t border-primary/10 pt-4">
             <p className="text-[10px] text-muted-foreground max-w-sm">
               Uses live backend monitoring routes for alert ingestion and observability polling. Wait for active simulation to receive payloads.
             </p>
             <div className="flex gap-2">
                <Button variant="outline" size="sm" className="text-xs h-8 rounded-full border-primary/30 text-primary hover:bg-primary/10">
                   View Last Payload
                </Button>
                <Button variant="default" size="sm" className="text-xs h-8 rounded-full bg-primary/80 hover:bg-primary text-primary-foreground shadow-[0_0_15px_rgba(139,92,246,0.3)]">
                   Send Test Event
                </Button>
             </div>
          </div>
        </CardContent>
      </Card>
    </div>
  );
};

export default IntegrationsPage;
