// frontend/app/flags/[id]/page.tsx
"use client";

import { useEffect, useState } from "react";
import { useParams, useRouter } from "next/navigation";
import { api, State, Observation } from "@/lib/api";
import { Button } from "@/components/ui/button";
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from "@/components/ui/card";
import { 
  ArrowLeft, 
  Save, 
  History, 
  Activity, 
  Split, 
  Users, 
  Target, 
  ShieldAlert,
  MessageSquare,
  TrendingUp
} from "lucide-react";
import { AnimatedToggle } from "@/components/ui/animated-toggle";
import { Slider } from "@/components/ui/slider";
import { Badge } from "@/components/ui/badge";
import { Progress } from "@/components/ui/progress";
import { AreaChart, Area, XAxis, YAxis, Tooltip, ResponsiveContainer, CartesianGrid } from "recharts";
import { cn } from "@/lib/utils";

export default function FlagDetailPage() {
  const params = useParams();
  const router = useRouter();
  const id = params.id as string;
  
  const [state, setState] = useState<State | null>(null);
  const [isOn, setIsOn] = useState(true);
  const [rollout, setRollout] = useState([25]);
  const [hasChanges, setHasChanges] = useState(false);
  const [message, setMessage] = useState<string>("");

  const fetchState = async () => {
    try {
      const s = await api.getState();
      setState(s);
      
      const history = s.history || [];
      const lastHistory = history.length > 0 ? history[history.length - 1] : null;
      const observation = lastHistory?.observation;
      
      if (observation && observation.feature_name === id) {
        setIsOn(!s.is_done);
        setRollout([observation.current_rollout_percentage]);
      }
    } catch (error) {
      console.error("Failed to fetch flag detail:", error);
    }
  };

  useEffect(() => {
    fetchState();
    const interval = setInterval(fetchState, 5000);
    return () => clearInterval(interval);
  }, [id]);

  const lastObs: Observation | undefined = state?.history?.[state.history.length - 1]?.observation;
  const isBackendFlag = lastObs?.feature_name === id;

  const handleRolloutChange = (val: number | number[] | readonly number[]) => {
    const newVal = typeof val === 'number' ? val : val[0];
    setRollout([newVal]);
    setHasChanges(true);
  };

  const handleToggle = (state: boolean) => {
    setIsOn(state);
    setHasChanges(true);
  };

  const handleSave = async () => {
    try {
      const targetRollout = rollout[0];
      let action_type = "MAINTAIN";
      if (!isOn) {
        action_type = "ROLLBACK";
      } else if (targetRollout === 100) {
        action_type = "FULL_ROLLOUT";
      } else if (targetRollout === 0) {
        action_type = "HALT_ROLLOUT";
      } else if (targetRollout > (lastObs?.current_rollout_percentage ?? 0)) {
        action_type = "INCREASE_ROLLOUT";
      } else {
        action_type = "DECREASE_ROLLOUT";
      }

      await api.step({
        action_type,
        target_percentage: targetRollout,
        reason: `Manual override via dashboard for ${id}`
      });
      
      setHasChanges(false);
      fetchState();
    } catch (error) {
      console.error("Failed to save changes:", error);
    }
  };

  return (
    <div className="flex-1 space-y-6 p-8">
      <div className="flex items-center justify-between">
        <div className="flex items-center gap-4">
          <Button variant="ghost" size="icon" onClick={() => router.push("/flags")} className="rounded-full hover:bg-muted">
            <ArrowLeft className="h-5 w-5" />
          </Button>
          <div>
            <div className="flex items-center gap-3">
              <h1 className="text-3xl font-bold tracking-tight capitalize">{id.replace(/-/g, ' ')}</h1>
              <Badge variant={isOn ? "default" : "secondary"} className="rounded-full">
                {isOn ? "Active" : "Inactive"}
              </Badge>
            </div>
            <p className="text-muted-foreground mt-1 text-sm font-mono">{id}</p>
          </div>
        </div>
        
        <div className="flex items-center gap-3">
          {hasChanges && (
            <Button onClick={handleSave} className="rounded-full bg-primary shadow-md shadow-primary/20">
              <Save className="mr-2 h-4 w-4" />
              Save Changes
            </Button>
          )}
        </div>
      </div>

      <div className="grid gap-6 md:grid-cols-3">
        <div className="md:col-span-2 space-y-6">
          {/* Mission Progress Section */}
          {isBackendFlag && lastObs?.mission_name && (
            <Card className="border-primary/20 bg-primary/5 shadow-none">
              <CardHeader className="pb-2">
                <div className="flex items-center justify-between">
                  <CardTitle className="text-sm font-bold flex items-center gap-2">
                    <Target className="h-4 w-4" />
                    Active Mission: {lastObs.mission_name}
                  </CardTitle>
                  <Badge variant="outline" className="bg-primary/10 border-primary/20">
                    Phase {lastObs.phase_index !== undefined ? lastObs.phase_index + 1 : 0}
                  </Badge>
                </div>
              </CardHeader>
              <CardContent className="space-y-4">
                <div className="flex items-center justify-between text-xs mb-1">
                   <span>{lastObs.current_phase}</span>
                   <span>{((lastObs.phase_progress ?? 0) * 100).toFixed(0)}%</span>
                </div>
                <Progress value={(lastObs.phase_progress ?? 0) * 100} className="h-2" />
                <div className="flex flex-wrap gap-2 pt-2">
                   {lastObs.phase_objectives?.map((obj, i) => (
                     <Badge key={i} variant="outline" className="text-[10px] bg-background/50">{obj}</Badge>
                   ))}
                </div>
              </CardContent>
            </Card>
          )}

          <Card className="border-border/50 bg-card/50 backdrop-blur-sm shadow-sm">
            <CardHeader className="flex flex-row items-center justify-between">
              <div>
                <CardTitle>Rollout Control</CardTitle>
                <CardDescription>Manage how this flag is evaluated.</CardDescription>
              </div>
              <AnimatedToggle isOn={isOn} onToggle={handleToggle} />
            </CardHeader>
            <CardContent className="space-y-8">
              <div className="space-y-4">
                <div className="flex items-center justify-between">
                  <span className="text-2xl font-bold text-primary">{rollout[0]}%</span>
                </div>
                <Slider 
                  value={rollout} 
                  onValueChange={handleRolloutChange} 
                  max={100} 
                  step={1} 
                  disabled={!isOn}
                  className="py-4"
                />
              </div>
            </CardContent>
          </Card>

          <Card className="border-border/50 bg-card/50 backdrop-blur-sm shadow-sm">
            <CardHeader>
              <CardTitle>Live Performance</CardTitle>
            </CardHeader>
            <CardContent>
              <div className="h-[250px] w-full">
                <ResponsiveContainer width="100%" height="100%">
                  <AreaChart data={state?.history?.map(h => ({ 
                    time: h.observation?.time_step, 
                    latency: h.observation?.latency_p99_ms,
                    errors: (h.observation?.error_rate ?? 0) * 100
                  })) || []}>
                    <CartesianGrid strokeDasharray="3 3" vertical={false} stroke="var(--border)" opacity={0.5} />
                    <XAxis dataKey="time" hide />
                    <YAxis fontSize={10} axisLine={false} tickLine={false} />
                    <Tooltip />
                    <Area type="monotone" dataKey="latency" stroke="var(--primary)" fill="var(--primary)" fillOpacity={0.1} />
                  </AreaChart>
                </ResponsiveContainer>
              </div>
            </CardContent>
          </Card>
        </div>

        <div className="space-y-6">
          {/* Stakeholder Sentiment Summary */}
          {isBackendFlag && (
             <Card className="border-border/50 bg-card/50">
                <CardHeader className="pb-2">
                  <CardTitle className="text-sm font-medium flex items-center gap-2">
                    <Users className="h-4 w-4" />
                    Stakeholder Sentiment
                  </CardTitle>
                </CardHeader>
                <CardContent className="space-y-4">
                   {[
                     { name: "DevOps", score: lastObs?.stakeholder_devops_sentiment },
                     { name: "Product", score: lastObs?.stakeholder_product_sentiment },
                     { name: "CS", score: lastObs?.stakeholder_customer_sentiment }
                   ].map(s => (
                     <div key={s.name} className="space-y-1">
                        <div className="flex items-center justify-between text-[10px] uppercase font-bold text-muted-foreground">
                           <span>{s.name}</span>
                           <span className={cn((s.score ?? 0) > 0 ? "text-green-600" : "text-red-600")}>
                             {(s.score ?? 0).toFixed(2)}
                           </span>
                        </div>
                        <Progress value={((s.score ?? 0) + 1) * 50} className={cn("h-1", (s.score ?? 0) < 0 && "[&>div]:bg-destructive")} />
                     </div>
                   ))}
                </CardContent>
             </Card>
          )}

          <Card className="border-border/50 bg-card/50">
            <CardHeader className="pb-2">
              <CardTitle className="text-sm font-medium flex items-center gap-2">
                <ShieldAlert className="h-4 w-4 text-amber-500" />
                Risk Profile
              </CardTitle>
            </CardHeader>
            <CardContent className="space-y-4">
              <div className="flex items-center justify-between text-sm">
                <span className="text-muted-foreground">Pattern Risk</span>
                <span className="font-bold">{((lastObs?.extra_context?.pattern_risk ?? 0) * 100).toFixed(0)}%</span>
              </div>
              <Progress value={(lastObs?.extra_context?.pattern_risk ?? 0) * 100} className="h-1.5" />
              
              <div className="pt-4 border-t mt-4">
                <div className="flex items-center justify-between text-xs mb-2">
                  <span className="text-muted-foreground font-medium uppercase">Active Anomaly</span>
                  {lastObs?.extra_context?.anomaly?.is_anomaly ? (
                    <Badge variant="destructive" className="h-4 text-[9px]">CRITICAL</Badge>
                  ) : (
                    <Badge variant="outline" className="h-4 text-[9px] text-green-600 border-green-200">CLEAR</Badge>
                  )}
                </div>
              </div>
            </CardContent>
          </Card>

          <Card className="border-border/50 bg-card/50">
            <CardHeader className="pb-2">
              <CardTitle className="text-sm font-medium flex items-center gap-2">
                <TrendingUp className="h-4 w-4 text-green-500" />
                Benchmark
              </CardTitle>
            </CardHeader>
            <CardContent>
               <div className="text-center py-2">
                  <span className="text-3xl font-bold">{(lastObs?.extra_context?.benchmarking?.percentile * 100 || 0).toFixed(0)}th</span>
                  <p className="text-[10px] text-muted-foreground mt-1">Percentile Performance</p>
               </div>
            </CardContent>
          </Card>
        </div>
      </div>
    </div>
  );
}
